### PR TITLE
feature : Add hmac validation to IPN controller

### PR DESCRIPTION
**Pull request by Francois-Gomis**
https://linear.app/almapay/issue/ECOM-1823/[-magento1]-add-hmac-verification-on-ipn

Add verification signature in IPN action. 

How to test :
Must be tested in QA env ! 
Create a product and set Alma **test & live** key
Create a paiement and check ipn logs in Alma dashbord

Use this payment id to check ipn on route : 
{baseURL}alma/payment/ipn?pid={paymentId}
without signature header and with a bad signature.

No unit test on magento 1 for the moment.

### CODE REVIEW

**Review by Benjamin-Freoua-Alma — APPROVED** [review]

File: src/app/code/community/Alma/Installments/Helper/Payment.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            };
+            return $signature;
+        }
+    }
+
+    /**
+     * @return void
+     * @throws Alma_installments_Model_Exceptions_PaymentException
+     */
+    public function checkSignature($almaPaymentId)
```
Comment (webaaz) [inline src/app/code/community/Alma/Installments/Helper/Payment.php:72]: parameters are not defined in PHPDOC

File: src/app/code/community/Alma/Installments/Helper/Payment.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    public function __construct()
+    {
+        $this->logger = Mage::helper('alma/Logger')->getLogger();
+    }
+
+    /**
+     * @return string
+     * @throws Alma_installments_Model_Exceptions_PaymentException
+     */
+    private function getHeaderSignature()
```
Comment (webaaz) [inline src/app/code/community/Alma/Installments/Helper/Payment.php:50]: feature is not explained in PHPDOC

File: src/app/code/community/Alma/Installments/Helper/Payment.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ *
+ * @author    Alma / Nabla SAS <contact@getalma.eu>
+ * @copyright 2024 Alma / Nabla SAS
+ * @license   https://opensource.org/licenses/MIT The MIT License
+ *
+ */
+
+class Alma_Installments_Helper_Payment extends Mage_Core_Helper_Abstract
```
Comment (webaaz) [inline src/app/code/community/Alma/Installments/Helper/Payment.php:30]: Class features are not explained in PHPDOC

File: src/app/code/community/Alma/Installments/Helper/Payment.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+     * @throws Alma_installments_Model_Exceptions_PaymentException
+     */
+    public function checkSignature($almaPaymentId)
+    {
+        if (!$this->isHmacValidated($almaPaymentId, $this->getApiKey(), $this->getHeaderSignature())) {
+            throw new Alma_installments_Model_Exceptions_PaymentException('Invalid signature');
+        }
+    }
+
+    private function isHmacValidated($almaPaymentId, $apiKey, $signature)
```
Comment (webaaz) [inline src/app/code/community/Alma/Installments/Helper/Payment.php:87]: feature is not explained in PHPDOC

File: src/app/code/community/Alma/Installments/Helper/Payment.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        return is_string($almaPaymentId) &&
+            is_string($apiKey) &&
+            hash_hmac('sha256', $almaPaymentId, $apiKey) === $signature;
+    }
+
+    /**
+     * @return string
+     * @throws Alma_installments_Model_Exceptions_PaymentException
+     */
+    private function getApiKey()
```
Comment (webaaz) [inline src/app/code/community/Alma/Installments/Helper/Payment.php:100]: feature is not explained in PHPDOC

File: src/app/code/community/Alma/Installments/Model/Exceptions/PaymentException.php [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+class Alma_installments_Model_Exceptions_PaymentException extends \Mage_Core_Exception
+{
+
+}
```
Comment (webaaz) [inline src/app/code/community/Alma/Installments/Model/Exceptions/PaymentException.php]: Empty ending line is not present on file

File: src/app/code/community/Alma/Installments/controllers/PaymentController.php [hunk 7]
```diff
@@ -194,9 +194,26 @@ public function returnAction()
 
     public function ipnAction()
```
Comment (webaaz) [inline src/app/code/community/Alma/Installments/controllers/PaymentController.php:225]: PHPDOC is missing, feature is not explained!

File: src/app/code/community/Alma/Installments/Helper/Payment.php (hunk 3 above)
Comment (Francois-Gomis) [reply src/app/code/community/Alma/Installments/Helper/Payment.php:30]: I hate Class doc bloc

**Pull request merged**